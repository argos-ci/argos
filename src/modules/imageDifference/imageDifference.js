import imageDifferenceRaw from 'image-difference'

// Generate an image diff result.
//
// We have been benchmarking other libraries, they are way too slow
// https://github.com/mapbox/pixelmatch
// https://github.com/gemini-testing/looks-same
//
// Some documentation
// https://en.wikipedia.org/wiki/Color_difference
async function imageDifference({
  compareScreenshotPath,
  baseScreenshotPath,
  diffResultPath,
  fuzz = 30 ** 2, // Small threshold to remove imperceptible changes.
}) {
  const difference = await imageDifferenceRaw({
    actualFilename: compareScreenshotPath,
    expectedFilename: baseScreenshotPath,
    diffFilename: diffResultPath,
    fuzz,
  })

  const score = difference.value / (difference.width * difference.height)

  return {
    score: score < 0.00003 ? 0 : score,
    pixels: difference.value,
    scoreRaw: score,
  }
}

export default imageDifference
