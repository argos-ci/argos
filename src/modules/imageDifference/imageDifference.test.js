import path from 'path'
import imageDifference from './imageDifference'

describe('imageDifference', () => {
  it('simple', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/simple/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/simple/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/simple/diff_tmp.png'),
      fuzz: 900,
    })

    expect(result.score).toBeCloseTo(0.306, 2)
    expect(result.pixels).toBeCloseTo(501e3, -3)
  })

  it('simple with enough fuzz', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/simple/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/simple/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/simple/diff_tmp.png'),
      fuzz: 70 ** 2,
    })

    expect(result.score).toBe(0)
    expect(result).toMatchSnapshot()
  })

  it('alphaBackground', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/alphaBackground/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/alphaBackground/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/alphaBackground/diff_tmp.png'),
    })

    expect(result.score).toBe(0)
    expect(result).toMatchSnapshot()
  })

  it('boxShadow', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/boxShadow/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/boxShadow/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/boxShadow/diff_tmp.png'),
    })

    expect(result.score).toBe(0)
    expect(result).toMatchSnapshot()
  })

  it('border', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/border/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/border/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/border/diff_tmp.png'),
    })

    expect(result.score).toBe(0)
  })

  it('fontAliasing', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/fontAliasing/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/fontAliasing/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/fontAliasing/diff_tmp.png'),
    })

    expect(result.score).toBeCloseTo(0, 2)
  })

  it('imageCompression', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/imageCompression/diff_tmp.png'),
    })

    expect(result.score).toBe(0)
    expect(result).toMatchSnapshot()
  })

  it('imageCompression2', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression2/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression2/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/imageCompression2/diff_tmp.png'),
    })

    expect(result).toMatchSnapshot()
  })

  it('imageCompression3', async () => {
    const result = await imageDifference({
      compareScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression3/compare.png'),
      baseScreenshotPath: path.join(__dirname, '__fixtures__/imageCompression3/base.png'),
      diffResultPath: path.join(__dirname, '__fixtures__/imageCompression3/diff_tmp.png'),
    })

    expect(result.pixels).toBeCloseTo(35, -1)
    expect(result.score).toBeCloseTo(0, 3)
  })
})
