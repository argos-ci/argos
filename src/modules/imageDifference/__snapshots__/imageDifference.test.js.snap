// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageDifference alphaBackground 1`] = `
Object {
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
}
`;

exports[`imageDifference boxShadow 1`] = `
Object {
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
}
`;

exports[`imageDifference imageCompression 1`] = `
Object {
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
}
`;

exports[`imageDifference imageCompression2 1`] = `
Object {
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
}
`;

exports[`imageDifference simple with enough fuzz 1`] = `
Object {
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
}
`;
